### ISSUE TITLE
WMT Jax workload does not converge and hangs

### ISSUE THREAD

**Issue by runame** [mlcommons/algorithmic-efficiency#105]
The WMT Jax workload does not converge in terms of train/validation loss and hangs mid-training (the job keeps running, but no logging messages are printed anymore).

## Description

See above. Notably, the workload worked correctly on January 11, 2022.

## Steps to Reproduce

Run WMT Jax workload using the standard instructions.

## Source or Possible Fix

No idea regarding the source, but the bug had to be introduced in one of the refactors, since the workload worked correctly before.

**Comment by runame** [comment]
Update: The cause of the lack of convergence is a typo in `model_fn()`, which leads to uncomplete inputs to the forward pass; see [this commit](https://github.com/runame/algorithmic-efficiency/commit/86bd8673f85c64f02b13dc4138e90c6b2200ee55) for the fix.

However, the training process still gets stuck at seemingly random steps. I verified that it does not hang during `data_selection()`, but in `update_params()`.

**Comment by runame** [comment]
I think the hanging issue could be caused by [this](https://github.com/google/jax/issues/10969). Updating jax and jaxlib to >= 0.3.14 should then solve the issue. However, after the update I get an error:
```
jax._src.traceback_util.UnfilteredStackTrace: jaxlib.xla_extension.XlaRuntimeError: INTERNAL: external/org_tensorflow/tensorflow/compiler/xla/service/gpu/nccl_utils.cc:245: NCCL operation ncclCommInitRank(comm.get(), nranks, id, rank) failed: unhandled cuda error: while running replica 0 and partition 0 of a replicated computation (other replicas may have failed as well).
```
[Here](https://github.com/google/jax/issues/10461) it is suggested that this might be caused by some mismatch of CUDA versions, but I couldn't identify anything like this in my setup. Would be interesting to know if you can reproduce this error after updating to the most recent jax and jaxlib versions or if the update fixes the hanging issue, @priyakasimbeg.

**Comment by runame** [comment]
Small update: as discussed in the last PyTorch engineering meeting I have verified that the workload does not hang on a single GPU.

**Comment by priyakasimbeg** [comment]
Ran this on updated main branch, on the VM with 8 GPUs, with Jax upgraded to and it still hangs :(.

JAX versions
```
jax==0.3.4                                                                                                                                                                         
jaxlib==0.3.2      
```
Note that I got an import error with Jax==0.3.16, so I tested with a higher version as recommended [here](https://stackoverflow.com/questions/72109033/cannot-import-name-isin-python).0

Output stuck after ~2h:
```
kasimbeg_google_com@in-to-win:~/algorithmic-efficiency$ python3 submission_runner.py --framework=jax --workload=wmt --submission_path=reference_submissions/wmt/wmt_jax/submi
ssion.py --tuning_search_space=reference_submissions/wmt/tuning_search_space.json --num_tuning_trials=1                                                                            
I0901 06:12:22.535927 139692803192640 xla_bridge.py:253] Unable to initialize backend 'tpu_driver': NOT_FOUND: Unable to find driver in registry given worker:                     
I0901 06:12:23.309264 139692803192640 xla_bridge.py:253] Unable to initialize backend 'tpu': INVALID_ARGUMENT: TpuPlatform is not available.                                       
I0901 06:12:24.374460 139692803192640 submission_runner.py:306] --- Tuning run 1/1 ---                                                                                             
I0901 06:12:24.489689 139692803192640 submission_runner.py:175] Initializing dataset.                                                                                              
I0901 06:12:24.492584 139692803192640 dataset_info.py:358] Load dataset info from /home/kasimbeg_google_com/tensorflow_datasets/wmt17_translate/de-en/1.0.0                        
I0901 06:12:24.497173 139692803192640 logging_logger.py:36] Constructing tf.data.Dataset wmt17_translate for split train, from /home/kasimbeg_google_com/tensorflow_datasets/wmt17_
translate/de-en/1.0.0                                                                                                                                                              
I0901 06:12:26.065483 139692803192640 submission_runner.py:182] Initializing model.                                                                                                
I0901 06:12:43.911551 139692803192640 submission_runner.py:185] Initializing optimizer.                                                                                            
I0901 06:12:45.473097 139692803192640 submission_runner.py:203] Starting training loop.                                                                                            
^[[B^[[B^[[B^[[BI0901 06:13:21.511332 139692803192640 spec.py:248] Evaluating on the training split.                                                                               
I0901 06:13:21.513684 139692803192640 dataset_info.py:358] Load dataset info from /home/kasimbeg_google_com/tensorflow_datasets/wmt17_translate/de-en/1.0.0                        
I0901 06:13:21.516190 139692803192640 logging_logger.py:36] Constructing tf.data.Dataset wmt17_translate for split train[:3001], from /home/kasimbeg_google_com/tensorflow_datasets
/wmt17_translate/de-en/1.0.0                                                                                                                                                       
^[[B^[[BI0901 06:13:34.442862 139692803192640 workload.py:189] Translating evaluation dataset.                                                                                     
I0901 06:18:50.205260 139692803192640 spec.py:259] Evaluating on the validation split.                                                                                             
I0901 06:18:50.206756 139692803192640 dataset_info.py:358] Load dataset info from /home/kasimbeg_google_com/tensorflow_datasets/wmt14_translate/de-en/1.0.0                        
I0901 06:18:50.208446 139692803192640 logging_logger.py:36] Constructing tf.data.Dataset wmt14_translate for split validation, from /home/kasimbeg_google_com/tensorflow_datasets/w
mt14_translate/de-en/1.0.0                    
                                                                                                                                     
...

I0901 08:26:44.478028 139692803192640 spec.py:248] Evaluating on the training split.
I0901 08:26:48.514737 139692803192640 workload.py:189] Translating evaluation dataset.
I0901 08:29:52.024291 139692803192640 spec.py:259] Evaluating on the validation split.
I0901 08:29:55.718309 139692803192640 workload.py:189] Translating evaluation dataset.
I0901 08:32:46.466060 139692803192640 submission_runner.py:252] 8039.00s        7315    {'train/accuracy': 0.6145674586296082, 'train/loss': 2.040191650390625, 'train/bleu': 26.01
5435549761712, 'validation/accuracy': 0.6203957200050354, 'validation/loss': 1.9960803985595703, 'validation/bleu': 22.203508777229626}
```

**Comment by runame** [comment]
Maybe I'm missing something, but `jax==0.3.4` is an earlier version than `jax==0.3.14`, which introduced the potential fix. Also, I just saw that yesterday there was a new Jax release (`0.3.17`) which you could try -- I will also try this new release.

**Comment by runame** [comment]
Unfortunately, I get the same error with `jax==0.3.17` as with all versions `>=0.3.14`. I think it might be a compatibility issue with the CUDA version used by our cluster (11.5) and [NCCL 2.12.12](https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-12-12.html#rel_2-12-12), but I cannot verify it because I can't easily control the CUDA version installed on our compute nodes.

**Comment by runame** [comment]
Also, could you tell me the exact import error you saw with `jax==0.3.16`, @priyakasimbeg?

**Comment by priyakasimbeg** [comment]
Hi Runa, you are absolutely right, 3.4 is not an upgrade lol. I don't know why I thought I installed 3.14 even though I clearly wrote 3.4. Thanks for the tip, the import error was from flax and after upgrading it was fixed!
Running a workload now, will post update here later.

**Comment by runame** [comment]
Haha nice! So it might have really been a compatibility issue for me. Now let's see if the workload still hangs...

**Comment by priyakasimbeg** [comment]
Still running... :) 
Btw, @runame  which version of CUDA is on your VM?

**Comment by runame** [comment]
Ok cool, that sounds promising. My VM has CUDA 11.5 -- the [release notes of NCCL 2.12.12](https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-12-12.html#rel_2-12-12) seem to imply that it is not supported, but I was not completely sure.

**Comment by runame** [comment]
Great to hear this. What batch size did you use? Previously I ran the workload with batch size 128 and it achieved a score of about 6864.

@sourabh2k15 Did you also try just running it with `jax>=0.3.14`? Otherwise yes, also setting the env variable should be fine.

**Comment by priyakasimbeg** [comment]
Hm that is weird. I ran it with default params on the main branch, so 128 based on [wmt_jax/submission.py](https://github.com/mlcommons/algorithmic-efficiency/blob/ebe3b950e04304f663f03591c14b72b81d3af9a0/reference_submissions/wmt/wmt_jax/submission.py#L14). 
For reference, the command: `python3 submission_runner.py --framework=jax --workload=wmt --submission_path=reference_submissions/wmt/wmt_jax/submission.py --tuning_search_space=reference_submissions/wmt/tuning_search_space.json --num_tuning_trials=1`

I think @sourabh2k15's librispeech run did end up hanging after 12h even with the `XLA_PYTHON_CLIENT_MEM_FRACTION = 0.85`. He will test if the updated version of jax works for librispeech as well.

**Comment by runame** [comment]
Ok very weird, how many steps did it take it to finish the workload?

**Comment by runame** [comment]
The score is an orthogonal issue, so I'm closing this for now. Fixed by PR #123 and #128.
